### PR TITLE
Improve compiler detection on MacOS

### LINKED ISSUE #121292: "g++" and "gcc" could be clang on MacOS
### 🐛 Describe the bug

In my debugging of `torch.compile` not working on an old MacOS X (12.7 Monterey), I've stumbled upon the place in PyTorch which doesn't look right:

https://github.com/pytorch/pytorch/blob/099ff51d459462cccf83361f33b9a59ecf93842e/torch/_inductor/codecache.py#L919-L920

But `g++` and `gcc` mean downright clang in my Mac:
```
$ gcc --version
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: x86_64-apple-darwin21.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

Apple's clang doesn't work for me anyway for reasons that are beyond me, I've had more progress with conda's clang (though still unresolved), so the code above doesn't affect me directly, but it looks like a sharp edge that might get someone into trouble with some MacOS setup.

### Error logs

_No response_

### Minified repro

_No response_

### Versions

Collecting environment information...
PyTorch version: 2.1.2.post102
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.7 (x86_64)
GCC version: Could not collect
Clang version: 16.0.6                   <-- This is conda's clang, not Apple's.
CMake version: version 3.28.3
Libc version: N/A

Python version: 3.11.8 | packaged by conda-forge | (main, Feb 16 2024, 20:51:20) [Clang 16.0.6 ] (64-bit runtime)
Python platform: macOS-12.7-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz

Versions of relevant libraries:
[pip3] flake8==7.0.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.25.2
[pip3] torch==2.1.2.post102
[conda] libblas                   3.9.0              20_osx64_mkl    conda-forge
[conda] libcblas                  3.9.0              20_osx64_mkl    conda-forge
[conda] liblapack                 3.9.0              20_osx64_mkl    conda-forge
[conda] libtorch                  2.1.2           cpu_mkl_h1b58e0f_102    conda-forge
[conda] mkl                       2023.2.0         h54c2260_50500    conda-forge
[conda] mkl-service               2.4.0           py311h5fe9819_1    conda-forge
[conda] numpy                     1.25.2          py311hc44ba51_0    conda-forge
[conda] pytorch                   2.1.2           cpu_mkl_py311h961e2af_102    conda-forge

cc @malfet @albanD @frank-wei @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @chauhang

### PR DESCRIPTION
**Pull request by xuhancn**
By relying on `is_apple_clang` helper function rather than on compiler name (as `gcc` is clang on MacOS):
```
% which gcc; gcc -v
/usr/bin/gcc
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.3.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
But
```
% /opt/homebrew/bin/gcc-13 -v
Using built-in specs.
COLLECT_GCC=/opt/homebrew/bin/gcc-13
COLLECT_LTO_WRAPPER=/opt/homebrew/Cellar/gcc/13.2.0/bin/../libexec/gcc/aarch64-apple-darwin23/13/lto-wrapper
Target: aarch64-apple-darwin23
Configured with: ../configure --prefix=/opt/homebrew/opt/gcc --libdir=/opt/homebrew/opt/gcc/lib/gcc/current --disable-nls --enable-checking=release --with-gcc-major-version-only --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-13 --with-gmp=/opt/homebrew/opt/gmp --with-mpfr=/opt/homebrew/opt/mpfr --with-mpc=/opt/homebrew/opt/libmpc --with-isl=/opt/homebrew/opt/isl --with-zstd=/opt/homebrew/opt/zstd --with-pkgversion='Homebrew GCC 13.2.0' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --with-system-zlib --build=aarch64-apple-darwin23 --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 13.2.0 (Homebrew GCC 13.2.0) 
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -917,6 +917,10 @@ def install_gcc_via_conda() -> str:
 
 
 def is_gcc() -> bool:
+    # Mac OS not support real gcc, it is acturally a clang.
+    _IS_MACOS = sys.platform.startswith("darwin")
+    if _IS_MACOS:
+        return False
```
Comment (malfet) [inline torch/_inductor/codecache.py:923]: Why not that, to be aligned with other similar checks in that file?
```suggestion
    if sys.platform == "darwin":
        return False
```

**Comment by xuhancn** [comment]
Issue submiter was no reply more than two weeks. Let merge the PR firstly? @malfet

File: torch/_inductor/codecache.py [hunk 2]
```diff
@@ -917,20 +917,26 @@ def install_gcc_via_conda() -> str:
 
 
 def is_gcc() -> bool:
+    # Mac OS not support real gcc, it is acturally a clang.
+    if sys.platform == "darwin":
```
Comment (malfet) [inline torch/_inductor/codecache.py:921]: This feels like a bit more correct behavior to me (I've seen a few issues when users installed gcc on Apple platform from anaconda. Even Inductor still has a logic to do so
```suggestion
    if sys.platform == "darwin" and is_apple_clang():
```

**Review by malfet — APPROVED** [review]
Did 3 things to this PR:
 - Applied suggested change as assuming that MacOS can not have gcc installed is wrong (`brew install gcc` indeed installs one), see output of gcc clang wrapper  vs real gcc in the description
 - Removed claim that it fixes users problem (as issue reporter did not reply nor PR description contained sufficient explanation why one should assume this is the case)
 - Updated PR title/description to be more explicit on what it does

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**